internal/action/workspace.go

```diff
@@ -143,15 +144,62 @@ func NewWorkspaceResource(ctx context.Context, client *tfe.Client, workspaces []
 	ws.SSHKeyID = config.SSHKeyID
 	ws.WorkingDirectory = config.WorkingDirectory
 
+	if err := SetTags(ws, config.Tags); err != nil {
+		return nil, err
+	}
+
 	return ws, nil
 }
 
+type Tag string
+type Tags []Tag
+
 type TeamDataResource struct {
 	ForEach      map[string]TeamDataResource `json:"for_each,omitempty"`
 	Name         string                      `json:"name"`
 	Organization string                      `json:"organization"`
 }
 
+// SetTags adds workspace tags to the passed module
+func SetTags(module *tfeprovider.Workspace, tags map[string]Tags) error {
+	if len(tags) == 0 {
+		return nil
+	}
+
+	b, err := json.Marshal(tags)
+	if err != nil {
+		return fmt.Errorf("failed to marshal workspace tags: %s", err)
+	}
+
+	module.TagNames = fmt.Sprintf("${toset(lookup(%s, each.value.name, []))}", string(b))
```

Comment:
yep agreed, however we are using a for_each on the workspace resource. This is probably one of the use cases that pushed hashi to promote doing loops in the calling language vs using a for each in the tfcdk. I'm inclined to go with the lookup approach for now until unless we decide it's worth it to pull apart the workspace resource or move to the tfcdk.

Reply:
Right right. We're in an in-betweener spot because workspace names are user-supplied data. Worth a second look at cdktf for whether they do anything for sanitizing strings into resource names. It would be very hard to switch, though I guess we could include a state migration in the action.

Seems like setting locals to the concrete value, letting that get serialized later, and just referring to the local in a dynamic expression would be an improvement over rendering JSON strings. Or actually a requirement, it's only a coincidence that marshaling to JSON works here, since JSON arrays are a subset of allowed HCL array syntax.

Reply:
yeah we could do a local lookup, no technical problem there, thought about it but went with this approach as it works (i was also a little surprised there) and it was a bit of a shortcut. 

If we moved over to the tfcdk, I might be inclined to just delete the s3 state file, allow the action to recreate the state file and import the newly addressed state

Reply:
> thought about it but went with this approach as it works (i was also a little surprised there) and it was a bit of a shortcut.

This at least needs a code comment, since the code itself (marshal JSON) does not make it clear what the actual intent is (marshal HCL). 

> If we moved over to the tfcdk, I might be inclined to just delete the s3 state file, allow the action to recreate the state file and import the newly addressed state

Mmm yeah that's probably appropriate. Release as a new major, re-generate state.

Reply:
Sorry didnt mean to push for the json in hcl approach, I'm on board with the local lookup, I think it makes the config easier to debug